### PR TITLE
HDPI-6452: Caseworker upload document

### PR DESCRIPTION
**Pull request by scottstewart-sl**
### Jira link

See [HDPI-6452](https://tools.hmcts.net/jira/browse/HDPI-6452)

### Change description

Add a screen to ExUI to allow a caseworker to upload a document to a case.

image

confirmation

### Testing done

Manual testing locally and in preview. Unit tests added/updated for new code.

### Security Vulnerability Assessment ###

**CVE Suppression:** Are there any CVEs present in the codebase (either newly introduced or pre-existing) that are being intentionally suppressed or ignored by this commit?
  * [ ] Yes
  * [x] No

### Checklist

- [x] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/domain/documentupload/CaseworkerDocument.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class CaseworkerDocument {
+
+    @CCD(ignore = true)
+    private static final String PREFIX = "cwDoc";
+
+    @CCD(label = "Upload a document")
+    @JsonProperty(PREFIX + "_Document")
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/domain/documentupload/CaseworkerDocument.java:29]: I've had to implement the prefix mechanism this way because when I added it to the `@JsonUnwrapped` annotation on this field in `PcsCase`, the submission data from ExUI was not deserialised back into this structure for nested objects. I was able to reproduce this with a simple Jackson test, so for now I have taken this approach and will investigate a better alternative separately.

**Review by libanAbdirahman1 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/document/DocumentNameService.java [hunk 2]
```diff
@@ -58,14 +64,14 @@ private static String buildFilename(String filename, String extension, String pa
...
-    public String appendCounterClaimDocumentName(String originalFilename, ClaimEntity claim, UUID partyId) {
+    public String appendCounterClaimPostfix(String originalFilename, ClaimEntity claim, UUID partyId) {
         if (originalFilename == null) {
             return null;
         }
 
         String baseName = FilenameUtils.getBaseName(originalFilename);
         String extension = FilenameUtils.getExtension(originalFilename);
-        String partyLabel = getPartyLabel(claim, partyId);
+        String partyLabel = partyService.getPartyLabel(claim, partyId);
```
Comment (tvr-hmcts) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/document/DocumentNameService.java:74]: Minor:  All tests for the class appear to return a partyLabel so do not assert on a null being returned and therefore the baseName below being used.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/caseworker/CaseworkerDocumentService.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                    caseFileCategory = CaseFileCategory.STATEMENTS_OF_CASE;
+                    break;
+                }
+                case NONE_PREFIX: {
+                    documentFilename = documentNameService.appendPartyPostfix(documentFilename, mainClaim, partyId);
+                    caseFileCategory = documentService.mapDocumentTypeToCategory(documentEntity.getType()).orElse(null);
+                    break;
+                }
+                default: {
+                    log.warn("Unexpected related submission prefix: " + codeParts[0]);
```
Comment (tvr-hmcts) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/caseworker/CaseworkerDocumentService.java:97]: minor: perhaps a verify test on this edge case to ensure nothing else is invoked?

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/SubmitHandler.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import uk.gov.hmcts.reform.pcs.ccd.domain.PCSCase;
+import uk.gov.hmcts.reform.pcs.ccd.domain.State;
+import uk.gov.hmcts.reform.pcs.ccd.domain.documentupload.CaseworkerDocument;
+import uk.gov.hmcts.reform.pcs.ccd.entity.DocumentEntity;
+import uk.gov.hmcts.reform.pcs.ccd.service.caseworker.CaseworkerDocumentService;
+import uk.gov.hmcts.reform.pcs.ccd.util.AddressFormatter;
+
+@Component("caseworkerUploadDocumentSubmitHandler")
+@RequiredArgsConstructor
+public class SubmitHandler implements Submit<PCSCase, State> {
```
Comment (tvr-hmcts) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/SubmitHandler.java:18]: Think this one needs coverage of its own.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/StartHandler.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+    private static List<RelatedSubmission> getGenAppSubmissions(List<ListValue<GeneralApplication>> genApps) {
+        return genApps.stream()
+            .map(genAppListValue -> {
+                GeneralApplication genApp = genAppListValue.getValue();
+
+                String displayLabel = "General Application GA%d".formatted(genApp.getRank());
+
+                LocalDateTime submittedDate = genApp.getSubmittedOn();
+                if (submittedDate != null) {
```
Comment (tvr-hmcts) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/StartHandler.java]: The tests always have a submittedDate

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/StartHandler.java [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+        return counterClaims.stream()
+            .map(counterClaimEntity -> {
+                UUID counterclaimPartyId = counterClaimEntity.getParty().getId();
+                String partyLabel = partyService.getPartyLabel(pcsCaseEntity.getMainClaim(), counterclaimPartyId);
+
+                String displayLabel = "Counterclaim - %s".formatted(partyLabel);
+
+                LocalDateTime submittedDate = counterClaimEntity.getClaimSubmittedDate();
+                if (submittedDate != null) {
```
Comment (tvr-hmcts) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/StartHandler.java]: Again here the submittedDate is always not null within the tests.

**Review by tvr-hmcts — COMMENTED** [review]
Some minors with some coverage needed on the SubmitHandler

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/document/DocumentNameService.java (hunk 2 above)
Comment (scottstewart-sl) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/document/DocumentNameService.java:74]: I've added a scenario for that now in the parameterised test

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/caseworker/CaseworkerDocumentService.java (hunk 3 above)
Comment (scottstewart-sl) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/caseworker/CaseworkerDocumentService.java:97]: I didn't add a verify, but I added a test to assert that the filename is not altered and the category is not set

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/SubmitHandler.java (hunk 4 above)
Comment (scottstewart-sl) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/SubmitHandler.java:18]: I've added this now, as well as a test for the event itself just to check it is calling these handlers

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/StartHandler.java (hunk 5 above)
Comment (scottstewart-sl) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/StartHandler.java]: Actually a gen app would always have a submitted date if it was visible in the case data so I've removed this null check

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/StartHandler.java (hunk 6 above)
Comment (scottstewart-sl) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/caseworker/uploaddocument/StartHandler.java]: A counterclaim would always have a submitted date once it has been issued, although your comment did make me realise that I wasn't filtering out pending counterclaims which I have now fixed and added a test for.

**Review by tvr-hmcts — APPROVED** [review]
LGTM :)
